VisualPinball.Unity/VisualPinball.Unity/VPT/Mech/ScoreMotorApi.cs

```diff
@@ -0,0 +1,277 @@
+
+		public DeviceCoil ResetCoil;
+		public DeviceSwitch MotorRunningSwitch;
+		public DeviceSwitch MotorStepSwitch;
+
+		private int _degreesPerStep;
+		private float _degreesPerSecond;
+
+		private bool _running;
+
+		private float _time;
+		private int _pos;
+
+		private ScoreMotorMode _mode;
+		private string _id;
+		private int _increase;
+
+		private float _score;
+		private ScoreMotorResetCallback _resetCallback;
+
+		private float _points;
+		private ScoreMotorAddPointsCallback _addPointsCallback;
+
+		public IApiSwitch Switch(string deviceItem)
+		{
+			return deviceItem switch
+			{
+				ScoreMotorComponent.MotorRunningSwitchItem => MotorRunningSwitch,
+				ScoreMotorComponent.MotorStepSwitchItem => MotorStepSwitch,
+				_ => throw new ArgumentException($"Unknown switch \"{deviceItem}\". "
+					+ "Valid names are \"{ScoreReelDisplayComponent.MotorRunningSwitchItem}\", and "
+					+ "\"{ScoreReelDisplayComponent.MotorStepSwitchItem}\".")
+			};
+		}
+
+		internal ScoreMotorApi(GameObject go, Player player)
+		{
+			_scoreMotorComponent = go.GetComponentInChildren<ScoreMotorComponent>();
+			_player = player;
+
+			_degreesPerSecond = _scoreMotorComponent.Degrees / (_scoreMotorComponent.Duration / 1000f);
+			_degreesPerStep = _scoreMotorComponent.Degrees / _scoreMotorComponent.Steps;
+
+			_scoreMotorComponent.OnReset += HandleReset;
+			_scoreMotorComponent.OnAddPoints += HandleAddPoints;
+
+			_running = false;
+		}
+
+		void IApi.OnInit(BallManager ballManager)
+		{
+			MotorRunningSwitch = new DeviceSwitch(ScoreMotorComponent.MotorRunningSwitchItem, false, SwitchDefault.NormallyOpen, _player);
+			MotorStepSwitch = new DeviceSwitch(ScoreMotorComponent.MotorStepSwitchItem, true, SwitchDefault.NormallyOpen, _player);
+
+			Init?.Invoke(this, EventArgs.Empty);
+		}
+
+		private void HandleReset(object sender, ScoreMotorResetEventArgs e)
+		{
+			if (_running) {
+				Logger.Info($"already running (ignoring reset), id={e.Id}");
+				return;
+			}
+
+			if (e.Score == 0) {
+				Logger.Info($"score already 0 (ignoring reset), id={e.Id}");
+				e.Callback(0);
+				return;
+			}
+
+			Logger.Info($"reset, id={e.Id}, score={e.Score}");
+
+			_mode = ScoreMotorMode.Reset;
+			_id = e.Id;
+			_score = e.Score;
+			_increase = ScoreMotorComponent.MaxIncrease;
+			_resetCallback = e.Callback;
+
+			StartMotor();
+		}
+
+		private void HandleAddPoints(object sender, ScoreMotorAddPointsEventArgs e)
+		{
+			var increase = (int)
+				((e.Points % 1000000000 == 0) ? e.Points / 1000000000 :
+				 (e.Points % 100000000 == 0) ? e.Points / 100000000 :
+				 (e.Points % 10000000 == 0) ? e.Points / 10000000 :
+				 (e.Points % 1000000 == 0) ? e.Points / 1000000 :
+				 (e.Points % 100000 == 0) ? e.Points / 100000 :
+				 (e.Points % 10000 == 0) ? e.Points / 10000 :
+				 (e.Points % 1000 == 0) ? e.Points / 1000 :
+				 (e.Points % 100 == 0) ? e.Points / 100 :
+				 (e.Points % 10 == 0) ? e.Points / 10 :
+				 e.Points);

```

Comment:
I don't understand this block. You're basically removing zeros from the score to add. But that score can be something like 1500, no? In this case you'll get 1 and drop the 500?

Reply:
This goes back to the `MaxIncrease` question / comment.  The final value calculates as 1, 2, 3, 4, or 5. 

On an EM with a score motor, you could never score 1500. @Scottacus64 can confirm this.

Reply:
Yes only multiples of 1, 10, 100, 1000 etc.  There would be no way for scoring three 500's, or five 300's.

Reply:
Got it. Although in this case, this block could be solved more elegantly, no? E.g.

```cs
var increase = Math.Floor(e.Points / Math.Pow(10, Math.Floor(Math.Log10(e.Points))));
```

Reply:
I'm not good with all that fancy math 😄. 

Updated: 

`
var increase = (int)System.Math.Floor(points / System.Math.Pow(10, System.Math.Floor(System.Math.Log10(points))));
`